src/vos/tests/vts_mvcc.c

```diff
@@ -569,32 +617,59 @@ conflicting_rw_exec_one(struct io_test_args *arg, int i, int j, bool empty,
 			struct op *r, char *rp, daos_epoch_t re,
 			struct op *w, char *wp, daos_epoch_t we, bool same_tx)
 {
-	struct mvcc_arg	*mvcc_arg = arg->custom;
-	struct dtx_id	 rtx;
-	struct dtx_id	 wtx;
-	int		 expected_rrc = 0;
-	int		 expected_wrc = 0;
-	int		 nfailed = 0;
-	int		 rc;
+	struct mvcc_arg		*mvcc_arg = arg->custom;
+	struct tx_helper	*rtx;
+	struct tx_helper	*wtx;
+	struct tx_helper	 txh1 = {0};
+	struct tx_helper	 txh2 = {0};
+	int			 expected_rrc = 0;
+	int			 expected_wrc = 0;
+	int			 nfailed = 0;
+	int			 rc;
 
-#if 1 /* FIXME: Many "same TX" cases fail currently. */
-	if (same_tx)
-		goto out;
-#endif
 	if (is_excluded(empty, r, rp, re, w, wp, we, same_tx))
 		goto out;
 
+	/*
+	 * Figure out the expected read result, perform read, and verify the
+	 * result.
+	 */
+	if (r->o_rtype == R_E && !empty)
+		expected_rrc = -DER_EXIST;
+	else if (r->o_rtype == R_NE && empty)
+		expected_rrc = -DER_NONEXIST;
+
+	if (same_tx && expected_rrc != 0) {
+		/** Not a valid use case as conditional updates are split in the
+		 *  context of distributed transactions.  The conditional fetch
+		 *  would mean either the update doesn't execute or should abort
+		 *  the transaction if it returns -DER_EXIST
+		 */
+		goto out;
+	}
+
```

Comment:
An application may do:

    daos_tx_open(..., &tx);
    rc = daos_obj_update(tx, "if dkey nonexist", dkey1, ...);
    if (rc == -DER_EXIST) {
        /* Do something else and commit tx. */
        daos_obj_punch(tx, dkey1);
        daos_obj_update(tx, dkey2, ...);
        ...
    }
    daos_tx_commit(tx);

So I think this change should be reverted, moving expected_rrc and comment back to near the read function call.

Reply:
I don't understand the comment.  If that happens, the dtx code translates it into a fetch + update so we will never see this case here

Reply:
Sure, fine with me. Regarding your reply before being edited, I think that's a good point. vos must handle

    same_tx && r is a conditional read

properly, while

    same_tx && r is a conditional update

is still under debate.